File: p523m/check_subarray_sum.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {boolean}
 */

var checkSubarraySum = function (nums, k) {
  let sumMap = new Map(); // sum to index map
  sumMap.set(0, -1);
  let sum = 0;
  for (let i = 0; i < nums.length; ++i) {
    sum += nums[i];
    sum %= k;
    let j = sumMap.get(sum);
    if (j !== undefined) {
      if (i - j >= 2) return true;
    } else {
      sumMap.set(sum, i);
    }
  }
  return false;
};
// TEST
[
  [[0, 0], 1, true],
  [[23, 2, 4, 6, 7], 6, true],
  [[23, 2, 6, 4, 7], 6, true],
  [[23, 2, 6, 4, 7], 13, false],
  [[1], 2, false],
].forEach(([nums, k, expected]) => {
  const actual = checkSubarraySum(nums, k);
  console.log('There is subarray in', nums, 'sum up to n *', k, '? ->', actual);
  console.assert(actual === expected);
});

File: p523m/check_subarray_sum.py
from itertools import accumulate


class Solution:
    def checkSubarraySum(self, nums: list[int], k: int) -> bool:
        psum_mod = [n % k for n in accumulate(nums)]
        seen = {0: -1}
        for i, n in enumerate(psum_mod):
            if n in seen:
                if i - seen[n] >= 2:
                    return True
            else:
                seen[n] = i
        return False


# TESTS
for nums, k, expected in [
    ([0, 0], 1, True),
    ([23, 2, 4, 6, 7], 6, True),
    ([23, 2, 6, 4, 7], 6, True),
    ([23, 2, 6, 4, 7], 13, False),
    ([1], 2, False),
]:
    sol = Solution()
    actual = sol.checkSubarraySum(nums, k)
    print(
        nums,
        "has subarray of size at least two sum up to multiple of",
        k,
        "->",
        actual,
    )
    assert actual == expected
